### ISSUE TITLE
JAWT_DrawingSurface_Lock() failed

### ISSUE THREAD

**Issue by hannespernpeintner** [LWJGLX/lwjgl3-awt#22]
Hey guys,

I tried to use your lib to have my OpenGL stuff displayed in a canvas in my swing/awt gui directly. My project is somewhat complicated, so I can't provide you a working example currently, but I get the following exception when I try to initialize an AWTGLCanvas:

```
Caused by: java.awt.AWTException: JAWT_DrawingSurface_Lock() failed
	at org.lwjgl.opengl.awt.PlatformLinuxGLCanvas.lock(PlatformLinuxGLCanvas.java:102)
	at org.lwjgl.opengl.awt.PlatformLinuxGLCanvas.create(PlatformLinuxGLCanvas.java:113)
	at org.lwjgl.opengl.awt.AWTGLCanvas.beforeRender(AWTGLCanvas.java:55)
```

I cannot find out what the problem here is, do you have any ideas?

Thx!!
Hannes

**Comment by httpdigest** [comment]
Can you do a test with LWJGL 2 and its AWTGLCanvas, whether that works?

**Comment by hannespernpeintner** [comment]
Yes I could do that, but I don't know how to get a test project. I can't add lwjgl 2 to my project, since I'm using lwjgl 3 already and that would clash I guess. Copying the AWTGLCanvas over alone didn't work at first sight, because it uses several other classes I don't have, I would have to copy them all, this would be a lot of work..

Do you have any other idea here, are there any demo projects I can simply clone and run?

**Comment by httpdigest** [comment]
You simply have to create a single 10 lines demo with a JFrame and a single AWTGLCanvas in it. That should be a thing of 1 minute max. You obviously don't have to completely port your engine/game/app to LWJGL 2 for that. Just a single super-simple example/demo that creates a JFrame + AWTGLCanvas with LWJGL 2.

**Comment by hannespernpeintner** [comment]
Ah, sorry, I misunderstood you, I thought I should test the AWTGLCanvas implementation in LWJGL 2 with the utils, helper classes and context implementations of LWJGL 3.

So yea, I created a small test project and that works well. If you want to take a look, it's https://github.com/hannespernpeintner/lwjgl2-awt-example . Does that help anything?

**Comment by hannespernpeintner** [comment]
Okay, guys, maybe I am wasting your time, I tried the AWTTest example, completely independent of my project and it works like a charm. So I have to assume I'm holding it wrong. Any chance that I can work with the given error somehow? What could be the reason the lock is not successful?

**Comment by hannespernpeintner** [comment]
Maybe it helps when i describe a little bit more what i try to achieve: i have a glfw window for context creation and after it is created i run my own opengl thread for everything, this is also a command queue that takes render commands. I want to switch to awt now and i have trouble translating the example so that my own thread just finally renders to the frame buffer of the awt canvas.

**Comment by hannespernpeintner** [comment]
Oh wow, I think I found out what the problem was. I started "rendering" in the canvas after initialization, but before the canvas was added to the frame.

@httpdigest thanks for your effort...again and again :D

**Comment by hannespernpeintner** [comment]
Hmmm...as we're on the topic already. My latest attempt seems to be unstable somehow. I wanted to prevent using the EDT for my OpenGL rendering stuff so I tried to adjust the AWTTest slightly.

```
ExecutorService pool = Executors.newFixedThreadPool(1)
...

Runnable renderLoop = new Runnable() {
	public void run() {
		if (!canvas.isValid())
			return;
		canvas.render();
                pool.submit(this);
//                SwingUtilities.invokeLater(this);
	}
};
	pool.submit(renderLoop);
//        SwingUtilities.invokeLater(renderLoop);
```

When I resize the window a lot, the rendering somewhen stops because the thread gets parked. Is it even possible to move the rendering out of the EDT?

**Comment by hannespernpeintner** [comment]
Small update here, I managed to include the AWTCanvas exactly like in the example in my project. I can now render everything into it. However, seems that the thing is still somehow unstable. Maybe you can work with for example this output

> [xcb] Unknown sequence number while processing queue
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
java: ../../src/xcb_io.c:259: poll_for_event: Assertion `!xcb_xlib_threads_sequence_lost' failed.

that gets printed before the complete application is killed.

Excessive reshaping of the window also gets me segfaults, maybe this can help:

> 
Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
J 7229  sun.awt.X11.XlibWrapper.XMoveResizeWindow(JJIIII)V java.desktop@11.0.4 (0 bytes) @ 0x00007f26c2de4023 [0x00007f26c2de3fc0+0x0000000000000063]
J 7342 c1 sun.awt.X11.XBaseWindow.xSetBounds(IIII)V java.desktop@11.0.4 (119 bytes) @ 0x00007f26bc106b94 [0x00007f26bc105d80+0x0000000000000e14]
j  sun.awt.X11.XWindow.reshape(IIII)V+41 java.desktop@11.0.4
j  sun.awt.X11.XContentWindow.reshape(IIII)V+6 java.desktop@11.0.4
j  sun.awt.X11.XWindow.reshape(Ljava/awt/Rectangle;)V+17 java.desktop@11.0.4
j  sun.awt.X11.XContentWindow.reshape(Ljava/awt/Rectangle;)V+2 java.desktop@11.0.4
j  sun.awt.X11.XContentWindow.setContentBounds(Lsun/awt/X11/WindowDimensions;)V+107 java.desktop@11.0.4
j  sun.awt.X11.XDecoratedPeer.reconfigureContentWindow(Lsun/awt/X11/WindowDimensions;)V+21 java.desktop@11.0.4
j  sun.awt.X11.XDecoratedPeer.handleConfigureNotifyEvent(Lsun/awt/X11/XEvent;)V+532 java.desktop@11.0.4
J 6859 c1 sun.awt.X11.XBaseWindow.dispatchEvent(Lsun/awt/X11/XEvent;)V java.desktop@11.0.4 (298 bytes) @ 0x00007f26bbfedda4 [0x00007f26bbfecd40+0x0000000000001064]
J 6857 c1 sun.awt.X11.XBaseWindow.dispatchToWindow(Lsun/awt/X11/XEvent;)V java.desktop@11.0.4 (44 bytes) @ 0x00007f26bbfebe84 [0x00007f26bbfeb0a0+0x0000000000000de4]
J 6856 c1 sun.awt.X11.XToolkit.dispatchEvent(Lsun/awt/X11/XEvent;)V java.desktop@11.0.4 (191 bytes) @ 0x00007f26bbfe86e4 [0x00007f26bbfe70c0+0x0000000000001624]
j  sun.awt.X11.XToolkit.run(Z)V+341 java.desktop@11.0.4
j  sun.awt.X11.XToolkit.run()V+5 java.desktop@11.0.4
j  java.lang.Thread.run()V+11 java.base@11.0.4
v  ~StubRoutines::call_stub

I think I'm going to quit my little experiment here, if you want to know anything else, feel free to respond :)

**Comment by ZNixian** [comment]
Sorry for the very late reply, but I ended up chasing down a my own similar issue. In a NetBeans Platform application, I was embedding a canvas and got this same error.

If you follow it into C, [awt_DrawingSurface_Lock](https://github.com/openjdk/jdk/blob/e209693a37e49ba5/src/java.desktop/unix/native/libawt_xawt/awt/awt_DrawingSurface.c#L63) has the relevant logic. It returns an error if one of four conditions is true:
* The supplied drawing surface was null
* The object passed to `awt_GetDrawingSurface` doesn't implement `java.awt.Component`
* `awt_init_Display` hasn't been called, which should never happen as that's done by AWT
* The `Component.peer` field of the object passed to `awt_GetDrawingSurface` (your canvas) is null

Looking at `PlatformLinuxGLCanvas`, the first option looks impossible (and on other platforms it's probably the same). If you're extending `AWTGLCanvas` then the second option is also impossible. The third one really can't happen unless something has gone horribly wrong inside AWT, but the last option is very possible.

If you call `render` before `addNotify` was called, you'll probably get this error - `Canvas.addNotify` sets up the `peer` field.

(As a last little note that I also found while tracking this down, if `disposeCanvas` is called on a different thread to where you first called `render` then it trips an assertion on OpenJDK debug builds, and on a release build can cause various weird VM crashes.)
